hax/hax/handler.py

```diff
@@ -111,16 +115,35 @@ def update_process_failure(self, q: Queue,
                           states=[HAState(fid=state.fid,
                                           status=ServiceHealth.FAILED)],
                           reply_to=None))
+                if current_status not in (ServiceHealth.UNKNOWN,
+                                          ServiceHealth.OFFLINE):
+                    # We also need to account and report the failure of remote
+                    # Motr processes to this node's hax and motr processes.
+                    # When Consul reports a remote process failure, hax
+                    # confirms its current status from Consul KV and updates
+                    # the list of failed services and also adds it to the
+                    # broadcast list.
+                    self.consul.update_failed_svcs(
```

Comment:
Looks like you put online services to the local list also? So how does the check for `if state.fid in self.consul.failed_svcs:` above works - will it check for online services also?

BTW, imagine some large cluster configuration with thousands ios processes - do we really want to maintain them in this list? This does not look like a scalable solution. I'd suggest to maintain only confd processes in it.

Reply:
>Looks like you put online services to the local list also? So how does the check for if state.fid in self.consul.failed_svcs: above work - won't it check for the online services also?

We add the process fid on failure and remove it when the process comes online from the failed_svc lists. So, failed_svcs list will have only the failed process fids.

>BTW, imagine some large cluster configuration with thousands ios processes - do we really want to maintain them in this list? This does not look like a scalable solution. I'd suggest to maintain only confd processes in it.

Yes, I did think about it, it will grow but presently I don't have an alternative solution for this. We can revisit this at later point. I though rconfc or RM would check the ha states for processes and services but seems like this does not happen and RM still tries to connect. That is the reason we need to maintain the list of failed services. Ideally there should be no need for this if the ha state updates are percolated through all the modules.

>Also, why do we need to maintain it in hax instead of Consul?

Because we don't need to persist the list, it is dynamic and can remain in-memory. Also during shutdown Consul might not be accessible as confds are stopped (No cluster leader will be available) and hax will not be able to fetch the failure list.

Reply:
> Also during shutdown Consul might not be accessible as confds are stopped (No cluster leader will be available) and hax will not be able to fetch the failure list.

Consul leader and RC leader and different things. Although, RC leader might not be available if there are no confds - true, Consul leader still will be available and you still will be able to access Consul KV even if all confds are down.

Reply:
Ah yes, here we need not get leader session, which fails is RC leader is not present. But even then the time required to fetch the list from Consul, look up or update and fetch again from consul is really not needed and this information can be maintained in memory. Let me try to reduce the list to just have confds.

Reply:
I thought the status of processes is updated in Consul KV anyway, isn't it so? If so, it will be only reads from Consul KV then.

The idea of maintaining processes status in Consul KV and in hax memory is not sound right.

Reply:
Okay, it was mainly about performance but let me try and see if there are any performance issues.

Reply:
Is it changed in the latest patch to read processes status from Consul?

Reply:
Yes. Hax now does not maintain a failed services list and fetches the status from Consul.